locker.go

```diff
@@ -0,0 +1,58 @@
+package osquery
+
+import (
+	"context"
+	"fmt"
+	"time"
+)
+
+// locker uses go channels to create a lock mechanism. We use channels, and not the more common mutexes, because the
+// latter cannot be interrupted. This allows callers to timeout without blocking on the mutex.
+//
+// We need _some_ lock mechanism because the underlying thrift socket only allows a single actor at a time. If two
+// goroutines are trying to use the socket at the same time, they will get protocol errors.
+type locker struct {
+	c              chan struct{}
+	defaultTimeout time.Duration // Default wait time is used if context does not have a deadline
+	maxWait        time.Duration // Maximum time something is allowed to wait
+}
+
+func NewLocker(defaultTimeout time.Duration, maxWait time.Duration) *locker {
+	return &locker{
+		c:              make(chan struct{}, 1),
+		defaultTimeout: defaultTimeout,
+		maxWait:        maxWait,
+	}
+}
+
+// Lock attempts to lock l. It will wait for the longer of (ctx deadline | defaultTimeout) and maxWait.
+func (l *locker) Lock(ctx context.Context) error {
+	// Assume most callers have set a deadline on the context, and start this as being the max allowed wait time
+	wait := l.maxWait
+	timeoutError := "timeout after maximum of %s"
+
+	// If the caller has not set a deadline, use the default.
+	if _, ok := ctx.Deadline(); !ok {
+		wait = l.defaultTimeout
+		timeoutError = "timeout after %s"
+
+	}
+
+	// Block until we get the lock, the context is canceled, or we time out.
+	select {
+	case l.c <- struct{}{}:
+		// lock acquired
+		return nil
+	case <-ctx.Done():
+		// context has been canceled
+		return fmt.Errorf("context canceled: %w", ctx.Err())
+	case <-time.After(wait):
```

Comment:
nit: "The underlying Timer is not recovered by the garbage collector until the timer fires".  
Depending on wait time and that is configurable and how fast lock is acquired, could have few timers hanging around until they fire.

Reply:
Nice catch. How's this?